### PR TITLE
Improve typescript support

### PR DESCRIPTION
**Pull request by SergioMorchon**
Export full type params of the exported class.
Add typescript to devDependencies.
Generate .d.ts file.
Reference .d.ts file in package.json.

### CODE REVIEW

File: build/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+declare const ReconnectingWebsocket: (url: string, protocols?: string | string[], options?: {
```
Comment (pladaria) [inline build/index.d.ts]: Is this file autogenerated? if not, don't place it inside the build folder. Move it to ./type-definitions

File: build/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    constructor?: {
+        new (url: string, protocols?: string | string[]): WebSocket;
```
Comment (pladaria) [inline build/index.d.ts]: This could be any function, not only a websocket

File: build/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        prototype: WebSocket;
```
Comment (pladaria) [inline build/index.d.ts]: why?

File: index.ts [hunk 4]
```diff
@@ -1,4 +1,14 @@
-const getDefaultOptions = () => ({
+type Options = {
+    constructor?: typeof WebSocket;
```
Comment (pladaria) [inline index.ts]: Same here, this shouldn't be limited to WebSockets

File: index.ts [hunk 5]
```diff
@@ -43,7 +53,7 @@ const reassignEventListeners = (ws, oldWs, listeners) => {
 const ReconnectingWebsocket = function(
     url: string,
     protocols?: string|string[],
-    options: Object = {}
+    options = <Options>{}
```
Comment (pladaria) [inline index.ts]: Just curious, is there any difference between `options: Options = {}` and `options = <Options>{}` ?

File: build/index.d.ts (hunk 1 above)
Comment (SergioMorchon) [inline build/index.d.ts]: This file is autogenerated, by adding the declarations entry in tsconfig.json.

File: build/index.d.ts (hunk 2 above)
Comment (SergioMorchon) [inline build/index.d.ts]: I see, I'll change the constructor typing so I could be any instantiable function

File: index.ts (hunk 5 above)
Comment (SergioMorchon) [inline index.ts]: Not really... The first one you first declare the expected type and then try to assign the value to it (with the possible typing errors) and the second one just cast the object to the <type> and the options type will be inferred.

**Comment by pladaria** [comment]
Thank you!

**Pull request merged**